crates/lsp-ai/src/memory_backends/vector_store.rs

```diff
@@ -0,0 +1,1062 @@
+                                        SpatialSimilarity::dot(vec1, vec2).context("vector length mismatch when taking the dot product")? as f32
+                                    )
+                                }
+                                #[cfg(not(feature = "simsimd"))]
+                                {
+                                    OrderedFloat(dot_product(&vec1, &vec2))
+                                }
+                            }
+                            (StoredChunkVec::Binary(vec1), StoredChunkVec::Binary(vec2)) => {
+                                #[cfg(feature = "simsimd")]
+                                {
+                                    OrderedFloat(
+                                        embedding.len() as f32
+                                            - BinarySimilarity::hamming(vec1, vec2).context("vector length mismatch when taking the hamming distance")?
+                                                as f32,
+                                    )
+                                }
+                                #[cfg(not(feature = "simsimd"))]
+                                {
+                                    OrderedFloat(
+                                        (embedding.len() - hamming_distance(&vec1, &vec2)) as f32,
+                                    )
+                                }
+                            }
+                            _ => anyhow::bail!("mismatch between vector data types in search"),
+                        };
+                        if acc.is_empty() {
+                            acc.insert(score, chunk);
+                        } else if acc.first_key_value().unwrap().0 < &score {
+                            // We want to get limit + 1 here in case the limit is 1 and then we filter the chunk out later
+                            if acc.len() == find_limit + 1 {
+                                acc.pop_first();
+                            }
+                            acc.insert(score, chunk);
+                        }
+                    }
+                    Ok(acc)
+                })
+                .collect();
+        let mut top_results = BTreeMap::new();
+        for result in results? {
+            for (sub_result_score, sub_result_chunk) in result {
+                let sub_result_score = if rerank_top_k.is_some() {
+                    match &sub_result_chunk.vec {
+                        StoredChunkVec::Binary(b) => {
+                            // Convert binary vector to f32 vec
+                            let mut b_f32 = vec![];
+                            for byte in b {
+                                for i in 0..8 {
+                                    let x = byte >> (8 - i) & 1;
+                                    b_f32.push(x as f32);
+                                }
+                            }
+                            b_f32.truncate(embedding.len());
```

Comment:
Could be nice to create a function for this operation so that it is easier to have some tests, seems like the kind of thing that can cause issues 😄

Reply:
(same goes for the opposite transformation)

Reply:
Great idea I will do that!